rfcs/0000-expressive-diagnostics.md

```diff
@@ -0,0 +1,356 @@
+
+Externally, we expect to see the effects experienced in other communities:
+
+- better perception of Sway as a language
+- easier language learning curve
+- increased efficiency in resolving compiler errors and warnings.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+When using `forc` via command line Sway programmers will get detailed diagnostics consisting of the following elements:
+
+- _Level_: Error or Warning.
+- _Code_: Unique code of the diagnostic.
+- _Reason_: Short description of the diagnostic, not related to a specific issue that the programmer did. The _reason_ answers the question "Why is this an error or warning?" E.g., Because - "Constants cannot be shadowed". It points out the general language rule that was violated.
+- _Issue_: Short description of the concrete issue caused by the programmer. E.g., "Variable "X" shadows imported constant with the same name"
+- _Hints_: Detailed descriptions of the diagnostic. Hints are placed in the source code. They point to the problematic line of code but also to other places in code that can help to understand the issue.
+- _Help_: Additional friendly information, which is not related to a place in code, that helps understanding and solving the issue.
+
+![oopsies image not showing](../files/0010-expressive-diagnostics-cli.png)
+
+When using IDEs like VS Code Sway programmers will have the experience similar to one offered by the [Rust analyzer](https://rust-analyzer.github.io/).
+
+Popup in the editor will provide the _reason_, _issue_, _hints_, as well as the _help_, similar to this Rust example:
+
+![oopsies image not showing](../files/0010-expressive-diagnostics-vscode-hints.png)
+
+Programmer will have the option to open the full compiler diagnostic, getting the same output like when using `forc` from the command line, similar to this Rust example:
+
+![oopsies image not showing](../files/0010-expressive-diagnostics-vscode-full-diagnostic.png)
+
+## Wording guidelines
+
+To ensure consistency and apply best practices<sup>[5]</sup> the diagnostics will follow these guidelines:
+
+- _Reason_ will be short and point out the language constraint which was violated. It will _not_ finish in fullstop, to emphasize brevity.
+- _Issue_ will be short and point out the specific situation encountered in code. It will also _not_ finish in fullstop, to emphasize brevity.
+- CLI or IDE plugin might choose to combine the _reason_ and the _issue_ as shown at the image above.
+- _Reason_ and _issue_ are given in plain english language, free of, e.g., compiler jargon.
+- _Hints_ and _help_ are as well written in plain english, using proper punctuation and grammar rules.
+- _Hints_ and _help_ try to give as much of useful context as possible and to be as specific as possible.
```

Comment:
What would go in `help` that could't be expressed in `hints`? For example, in the `cannot be shadowed` example, it says that constants cannot be shadowed multiple times in slightly different ways. The longer the message, the harder it is to digest, so we should just say it once at the top, and have a hint about how to fix it.